oss-internship-2020/sapi_lodepng/examples/main_sandboxed.cc

```diff
@@ -0,0 +1,211 @@
+      image[4 * width * y + 4 * x + 2] = x | y;
+      image[4 * width * y + 4 * x + 3] = 255;
+    }
+  }
+
+  // encode the image
+  sapi::v::Array<unsigned char> sapi_image(image.data(), img_len);
+  sapi::v::UInt sapi_width(width), sapi_height(height);
+  std::string filename = "/output/out_generated1.png";
+  sapi::v::ConstCStr sapi_filename(filename.c_str());
+
+  sapi::StatusOr<unsigned int> result = api.lodepng_encode32_file(
+      sapi_filename.PtrBefore(), sapi_image.PtrBefore(), sapi_width.GetValue(),
+      sapi_height.GetValue());
+
+  assert(result.ok());
+  assert(!result.value());
+
+  // after the image has been encoded, decode it to check that the
+  // pixel values are the same
+
+  sapi::v::UInt sapi_width2, sapi_height2;
+  sapi::v::IntBase<unsigned char *> sapi_image_ptr(0);
+
+  result = api.lodepng_decode32_file(
+      sapi_image_ptr.PtrBoth(), sapi_width2.PtrBoth(), sapi_height2.PtrBoth(),
+      sapi_filename.PtrBefore());
+  assert(result.ok());
+  assert(!result.value());
+
+  assert(sapi_width2.GetValue() == width);
+  assert(sapi_height2.GetValue() == height);
+
+  // the pixels have been allocated inside the sandboxed process
+  // memory, so we need to transfer them to this process.
+  // Transferring the memory has the following steps:
+  // 1) define a RemotePtr variable that holds the memory location from
+  // the sandboxed process
+  // 2) define an array with the required length
+  // 3) set the remote pointer for the array to specify where the memory
+  // that will be transferred is located
+  // 4) transfer the memory to this process (this step is why we need
+  // the pointer and the length)
+  sapi::v::RemotePtr sapi_remote_out_ptr(sapi_image_ptr.GetValue());
+  sapi::v::Array<unsigned char> sapi_pixels(img_len);
+  sapi_pixels.SetRemote(sapi_remote_out_ptr.GetValue());
+
+  assert(sandbox.TransferFromSandboxee(&sapi_pixels).ok());
+
+  // after the memory has been transferred, we can access it
+  // using the GetData function
+  unsigned char *pixels_ptr = sapi_pixels.GetData();
+
+  // now, we can compare the values
+  for (size_t i = 0; i < img_len; ++i) {
+    assert(pixels_ptr[i] == image[i]);
+  }
+}
+
+void generate_two_steps(SapiLodepngSandbox &sandbox, LodepngApi &api) {
+  // generate the values
+  constexpr unsigned int width = 512, height = 512,
+                         img_len = width * height * 4;
+  std::vector<unsigned char> image(img_len);
+
+  for (int y = 0; y < height; ++y) {
+    for (int x = 0; x < width; ++x) {
+      image[4 * width * y + 4 * x + 0] = 255 * !(x & y);
+      image[4 * width * y + 4 * x + 1] = x ^ y;
+      image[4 * width * y + 4 * x + 2] = x | y;
+      image[4 * width * y + 4 * x + 3] = 255;
+    }
+  }
+
+  // encode the image into memory first
+  sapi::v::Array<unsigned char> sapi_image(image.data(), img_len);
+  sapi::v::UInt sapi_width(width), sapi_height(height);
+  std::string filename = "/output/out_generated2.png";
```

Comment:
As in the previous file, probably no need to have an additional variable just to encode filename. If you want to keep it, maybe make it const?